### PR TITLE
item-inventory

### PR DESCRIPTION
**Pull request by mradulrathore**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+*.dll
+*.so
+*.dylib
+*.vscode
+
+# Test binary, build with `go test -c`
+*.test
+
+# Output of the go coverage tool, specifically when used with LiteIDE
+*.out
```
Comment (10yogi) [inline .gitignore]: always have endline

File: constants/constants.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package constants
+
+const (
+	TaxRateForRAW                                    = 12.50
```
Comment (10yogi) [inline constants/constants.go]: RAWItemTaxRate ?

File: constants/constants.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+	ImportDuty                                       = 10.00
```
Comment (10yogi) [inline constants/constants.go]: use fractions

File: constants/constants.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+	ImportDutyLimit1                                 = 100
+	ImportDutyLimit2                                 = 200
+	SurchargeAmountForFinalCostUptoImportDutyLimit1  = 5
```
Comment (10yogi) [inline constants/constants.go]: abbr. can be used. read go naming conventions.

File: constants/constants.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+	SurchargeAmountForFinalCostUptoImportDutyLimit2  = 10
+	SurchargeRateForFinalCostExceedeImportDutyLimit2 = 5.00
```
Comment (10yogi) [inline constants/constants.go]: naming can be improved.

File: items/item.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
+package item
+
+import (
+	constant "application/constants"
```
Comment (10yogi) [inline items/item.go]: use pakcage naming like
eg.
"github.com/CDNA-Technologies/proto-gen/go/arise/tenant/v1"

File: items/item.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+	if item.Price < 0 {
+		return false, errors.New("price can not be negative")
+	}
+	if item.TypeItem != "raw" && item.TypeItem != "manufactured" && item.TypeItem != "imported" {
+		return false, errors.New("item type can only be raw, manufactured or imported")
+	}
+	return true, nil
+}
+
+func GetUserChoice() (string, error) {
```
Comment (10yogi) [inline items/item.go]: separation of concerns.

File: .gitignore (hunk 1 above)
Comment (mradulrathore) [reply .gitignore]: added a new line at the end.

File: constants/constants.go (hunk 2 above)
Comment (mradulrathore) [reply constants/constants.go]: okay! added proper constant and variable names

File: constants/constants.go (hunk 3 above)
Comment (mradulrathore) [reply constants/constants.go]: done!

File: constants/constants.go (hunk 4 above)
Comment (mradulrathore) [reply constants/constants.go]: added abbr. (ImportDutyLimit2SurchargeAmt)

File: constants/constants.go (hunk 5 above)
Comment (mradulrathore) [reply constants/constants.go]: improved the naming (ImportDutyLimit2SurchargeAmt)

File: items/item.go (hunk 6 above)
Comment (mradulrathore) [reply items/item.go]: done! (added github.com/mradulrathore/onboarding-assignments)

File: items/item.go (hunk 7 above)
Comment (mradulrathore) [reply items/item.go]: added view package

File: item/item_test.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+func TestCalculateTaxAndPrice(t *testing.T) {
+	var tests = []struct {
+		req                       Item
+		ExpectedSalesTaxLiability float64
+		err                       error
+	}{
+		// all item details provided
+		{
```
Comment (10yogi) [inline item/item_test.go]: add scenario field: brief desc

File: constant/constant.go [hunk 9]
```diff
@@ -0,0 +1,18 @@
+package constant
+
+const (
```
Comment (10yogi) [inline constant/constant.go]: move  it to proper domain

File: constant/constant.go [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+	ImportDutyLimit2SurchargeAmt         = 10
+	ExceedeImportDutyLimit2SurchargeRate = 0.05
+	ManufacturedItmTaxRate               = 0.125
+	ManufacturedItmExtraTaxRate          = 0.02 //Extra =ItemCost +12.5% Item Cost
+	Accept                               = "y"
+	Deny                                 = "n"
+	Raw                                  = "raw"
+	Manufactured                         = "manufactured"
+	Imported                             = "imported"
+)
```
Comment (10yogi) [inline constant/constant.go]: check whether we can use enum
lib: enumer

File: coverage.out [hunk 11]
```diff
@@ -0,0 +1 @@
+mode: set
```
Comment (10yogi) [inline coverage.out]: delete it

File: item/item.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+	"github.com/mradulrathore/onboarding-assignments/constant"
+)
+
+type Item struct {
+	Name              string
+	Price             float64
+	Quantity          int
+	Type              string
+	SalesTaxLiability float64
+	FinalPrice        float64
```
Comment (10yogi) [inline item/item.go]: Use function.
 Read about generated fields.

File: item/item.go [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+func (item *Item) CalculateRAWTax() error {
+	item.SalesTaxLiability = constant.RAWItmTaxRate * item.Price
+	return nil
+}
```
Comment (10yogi) [inline item/item.go]: don't expose

File: item/item.go [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+	item.FinalPrice = item.Price + item.SalesTaxLiability
+	return nil
+}
+
+func (item *Item) CalculateImportTax() error {
+	item.SalesTaxLiability = constant.ImportDuty * item.Price
+	return nil
+}
+
+func (item *Item) CalculateImportFinalPrice() error {
```
Comment (10yogi) [inline item/item.go]: check about order  & context of func or info

clean code

File: item/item.go [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+	SalesTaxLiability float64
+	FinalPrice        float64
+}
+
+func (item *Item) CalculateRAWTax() error {
+	item.SalesTaxLiability = constant.RAWItmTaxRate * item.Price
+	return nil
+}
+
+func (item *Item) CalculateRAWFinalPrice() error {
```
Comment (10yogi) [inline item/item.go]: don't add additional code for the future, prefer TODO comments

File: item/item.go [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+	item.FinalPrice = item.Price + item.SalesTaxLiability
+	return nil
+}
+
```
Comment (10yogi) [inline item/item.go]:
```suggestion

func (item *Item) GetTax() error {
	item.SalesTaxLiability = constant.RAWItmTaxRate * item.Price
	return nil
}

func (item *Item) GetImport() error {
	item.GetTax() + Price()
}

func (item *Item) GetFinalPrice() error {
	return item.Price + item.Get(Tax) + item.GetIport()
}
```

File: item/item.go [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+	}
+
+	return nil
+}
+
+func (item *Item) CalculateTaxAndPrice() error {
+	switch item.Type {
+	case constant.Raw:
+		//raw: 12.5% of the item cost
+		err := item.CalculateRAWTax()
```
Comment (10yogi) [inline item/item.go]: tips: read about error handling

File: item/item.go [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+		if err != nil {
+			log.Println(err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (item Item) GetItemDetails() error {
```
Comment (10yogi) [inline item/item.go]: don't have this.
add json tags,

automated process to add tags

File: item/item.go [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+	fmt.Printf("Item Name: %s \n", item.Name)
+	fmt.Printf("Item Price: %g \n", item.Price)
+	fmt.Printf("Item Quantity: %d \n", item.Quantity)
+	fmt.Printf("Item Type: %s \n", item.Type)
+	fmt.Printf("Item Type: %g \n", item.SalesTaxLiability)
+	fmt.Printf("Item Type: %g \n \n", item.FinalPrice)
+
+	return nil
+}
+
```
Comment (10yogi) [inline item/item.go]: NewItem( ....){
Validate().
}

File: item/item.go [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+			log.Println(err)
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+func (item *Item) ValidateItemDetails() error {
```
Comment (10yogi) [inline item/item.go]: Validate()

File: item/item.go [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+	if len(item.Type) == 0 {
```
Comment (10yogi) [inline item/item.go]: user item.Type != ""

File: item/item.go [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+		return custErr.InvalideItmType
+	}
```
Comment (10yogi) [inline item/item.go]: ozzovalidation

File: item/item.go [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+	if item.Quantity < 0 {
+		return custErr.NegativeQuantErr
```
Comment (10yogi) [inline item/item.go]: use ozzo validation

File: view/view.go [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+	return moreItems, nil
+}
+
+// validate whether userChoice is eiter Accept or Deny
+func ValidateConfirmation(userChoice string) error {
+
+	if userChoice != constant.Accept && userChoice != constant.Deny {
+		log.Println("enter either " + constant.Accept + " or " + constant.Deny)
+		return custErr.InvalidUsrChoice
+	}
```
Comment (10yogi) [inline view/view.go]: refactor code for contextual alignment.

File: main.go [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+	price    = flag.Float64("price", 0, "price of item")
+	quantity = flag.Int("quantity", 0, "quantity of item")
+	typeItem = flag.String("type", "", "type of item")
+)
+
+func main() {
+
+	flag.Parse()
+
+	log.Println("item name: ", *name)
```
Comment (10yogi) [inline main.go]: single line long

File: view/view.go [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+	_, err := fmt.Scanf("%s", &moreItems)
+	if err != nil {
+		log.Println(err)
+		return "", err
+	}
+
+	err = ValidateConfirmation(moreItems)
+
+	for err != nil {
+
```
Comment (10yogi) [inline view/view.go]: why?

File: view/view.go [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+	"fmt"
+	"log"
+
+	"github.com/mradulrathore/onboarding-assignments/constant"
+	custErr "github.com/mradulrathore/onboarding-assignments/error"
+	itm "github.com/mradulrathore/onboarding-assignments/item"
+)
+
+func GetUserChoice() (string, error) {
+
```
Comment (10yogi) [inline view/view.go]: why new line?

File: view/view.go [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		log.Println(err)
+		return "", err
+	}
+
+	err = ValidateConfirmation(moreItems)
+
+	for err != nil {
+
+		_, err = fmt.Scanf("%s", &moreItems)
```
Comment (10yogi) [inline view/view.go]: duplicate code.

use recursion or move outside.

File: view/view.go [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+			return "", err
+		}
+		err = ValidateConfirmation(moreItems)
+	}
+
+	return moreItems, nil
+}
+
+// validate whether userChoice is eiter Accept or Deny
+func ValidateConfirmation(userChoice string) error {
```
Comment (10yogi) [inline view/view.go]: don't expose

File: view/view_test.go [hunk 30]
```diff
@@ -0,0 +1,29 @@
+package view
+
+// cmd go test -coverprofile=coverage.out
+
+import (
+	"testing"
+
+	custErr "github.com/mradulrathore/onboarding-assignments/error"
```
Comment (10yogi) [inline view/view_test.go]: avoid it

File: coverage.out (hunk 11 above)
Comment (mradulrathore) [reply coverage.out]: done!

File: main.go (hunk 25 above)
Comment (mradulrathore) [reply main.go]: refactored the code.

File: constant/constant.go (hunk 9 above)
Comment (mradulrathore) [reply constant/constant.go]: moved the constant to item and view package

File: view/view_test.go (hunk 30 above)
Comment (mradulrathore) [reply view/view_test.go]: okay, made the changes.

File: item/item.go [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+	item.Type, err = enum.ItemTypeString(typeItem)
+	if err != nil {
+		return
+	}
+
+	err = item.validate()
+	return
+}
+
+// func checkNegativeValue(value interface{}) (err error) {
```
Comment (10yogi) [inline item/item.go]: prefer to not have commented code. or add a proper explanation why u commented.

File: view/view_test.go [hunk 32]
```diff
@@ -0,0 +1,120 @@
...
+		scenario: "all item details provided for imported",
+		req:      setInput(testImported),
+		err:      nil,
+	},
+		{
+			scenario: "invalid item type, exported",
+			req:      setInput(testInvalidItmType),
+			err:      invalidTypeErr,
+		},
+	}
```
Comment (10yogi) [inline view/view_test.go]: ? formatting

Comment (mradulrathore) [reply view/view_test.go]: formatted the code

File: item-inventory/item/constant.go [hunk 33]
```diff
@@ -0,0 +1,16 @@
...
+	ImportDuty                           = 0.100
+	ImportDutyLimit1                     = 100
+	ImportDutyLimit2                     = 200
+	ImportDutyLimit1SurchargeAmt         = 5
+	ImportDutyLimit2SurchargeAmt         = 10
+	ExceedeImportDutyLimit2SurchargeRate = 0.05
+	ManufacturedItmTaxRate               = 0.125
+	ManufacturedItmExtraTaxRate          = 0.02 //Extra =ItemCost +12.5% Item Cost
+	Raw                                  = "raw"
+	Manufactured                         = "manufactured"
```
Comment (10yogi) [inline item-inventory/item/constant.go]: not required

File: item-inventory/item/enum/item.go [hunk 34]
```diff
@@ -0,0 +1,11 @@
+package enum
+
+//go:generate enumer -type=ItemType
+
```
Comment (10yogi) [inline item-inventory/item/enum/item.go:4]: push generated code

File: item-inventory/item/enum/item.go [hunk 35]
```diff
@@ -0,0 +1,11 @@
...
+const (
+	Raw ItemType = iota
+	Manufactured
+	Imported
+)
+
+type ItemType int
```
Comment (10yogi) [inline item-inventory/item/enum/item.go]: move to top

File: item-inventory/item/item.go [hunk 36]
```diff
@@ -0,0 +1,95 @@
...
+)
+
+type Item struct {
+	Name     string        `json:"name"`
+	Price    float64       `json:"price"`
+	Quantity int           `json:"quantity"`
+	Type     enum.ItemType `json:"type"`
+}
+
+func New(name string, price float64, quantity int, typeItem string) (item Item, err error) {
```
Comment (10yogi) [inline item-inventory/item/item.go]: don't name return types.
fix this

File: item-inventory/item/item.go [hunk 37]
```diff
@@ -0,0 +1,95 @@
...
+func (item Item) validate() (err error) {
+	if item.Quantity < 0 {
+		err = NegativeQuantErr
+	}
+	if item.Price < 0 {
+		err = NegativePriceErr
+	}
+	return
+	//return validation.ValidateStruct(&item, validation.Field(&item.Quantity, validation.By(checkNegativeValue)))
+}
```
Comment (10yogi) [inline item-inventory/item/item.go]: add TODO:
remove comments

File: item-inventory/item/item.go [hunk 38]
```diff
@@ -0,0 +1,95 @@
...
+
+func (item Item) String() string {
+	return fmt.Sprintf("[%s, %g, %d,%s,%g,%g]", item.Name, item.Price, item.Quantity, item.Type.String(), item.getTax(), item.getEffectivePrice())
+}
```
Comment (10yogi) [inline item-inventory/item/item.go]: GetInvoice

GetInvoice()

GetEffectivePrice

GetTax()

File: item-inventory/item/error.go [hunk 39]
```diff
@@ -0,0 +1,10 @@
+package item
+
+import (
+	"errors"
+)
+
+var (
+	NegativeQuantErr = errors.New("quantity can not be negative")
+	NegativePriceErr = errors.New("price can not be negative")
+)
```
Comment (10yogi) [inline item-inventory/item/error.go]: errors.New(
remove it.

add context to error.

user place holders for context.

File: item-inventory/view/error.go [hunk 40]
```diff
@@ -0,0 +1,9 @@
+package view
+
+import (
+	"errors"
+)
+
+var (
+	InvalidUsrChoice = errors.New("enter either " + Accept + " or " + Deny)
```
Comment (10yogi) [inline item-inventory/view/error.go]: errors.New(
 
 try to add context.
 
 add place, holders

File: item-inventory/view/view.go [hunk 41]
```diff
@@ -0,0 +1,96 @@
...
+
+func Initialize() (err error) {
+	name, price, quantity, typeItem, err := getItem()
+	if err != nil {
+		return
+	}
+	item, err := itm.New(name, price, quantity, typeItem)
+
+	for err != nil {
+		log.Println(err.Error())
```
Comment (10yogi) [inline item-inventory/view/view.go:25]: no line b/w err

File: item-inventory/item/item_test.go [hunk 42]
```diff
@@ -0,0 +1,99 @@
...
+	"testing"
+
+	"github.com/mradulrathore/onboarding-assignments/item-inventory/item/enum"
+)
+
+//Doubt
+func TestNew(t *testing.T) {
+	_, emptyTypeErr := enum.ItemTypeString("")
+	_, invalidTypeErr := enum.ItemTypeString("exported")
+
```
Comment (10yogi) [inline item-inventory/item/item_test.go]: don't use

File: item-inventory/item/enum/item.go (hunk 34 above)
Comment (mradulrathore) [reply item-inventory/item/enum/item.go:4]: added Makefile and updated .gitignore file to include "enumer" generated code also

File: item-inventory/item/constant.go (hunk 33 above)
Comment (mradulrathore) [reply item-inventory/item/constant.go]: removed the unnecessary constants

File: item-inventory/item/enum/item.go (hunk 35 above)
Comment (mradulrathore) [reply item-inventory/item/enum/item.go]: done!

File: item-inventory/item/item.go (hunk 36 above)
Comment (mradulrathore) [reply item-inventory/item/item.go]: fixed!

File: item-inventory/item/item.go (hunk 38 above)
Comment (mradulrathore) [reply item-inventory/item/item.go]: refactored the name and exposed the methods

File: item-inventory/item/item.go (hunk 37 above)
Comment (mradulrathore) [reply item-inventory/item/item.go]: added ooze validation!

**Review by 10yogi — APPROVED** [review]

**Review by 10yogi — APPROVED** [review]

**Pull request merged**